File: src/KoalaWiki/Services/MenuService.cs
using FastService;
using KoalaWiki.Core.DataAccess;
using KoalaWiki.Dto;
using KoalaWiki.Infrastructure;
using MapsterMapper;
using Microsoft.AspNetCore.Authorization;
using Microsoft.EntityFrameworkCore;

namespace KoalaWiki.Services;

/// <summary>
/// 菜单管理服务
/// </summary>
[Tags("菜单管理")]
[Route("/api/Menu")]
[Filter(typeof(ResultFilter))]
public class MenuService(
    IKoalaWikiContext dbContext,
    IMapper mapper,
    ILogger<MenuService> logger,
    IUserContext userContext) : FastApi
{
    /// <summary>
    /// 获取当前用户的菜单
    /// </summary>
    /// <returns>用户菜单</returns>
    [EndpointSummary("获取当前用户的菜单")]
    [Authorize]
    public async Task<UserMenuDto> GetUserMenuAsync()
    {
        try
        {
            var userId = userContext.CurrentUserId;
            if (string.IsNullOrEmpty(userId))
            {
                throw new UnauthorizedAccessException("用户未登录");
            }

            // 获取用户信息
            var user = await dbContext.Users
                .AsNoTracking()
                .FirstOrDefaultAsync(u => u.Id == userId);

            if (user == null)
            {
                throw new ArgumentException("用户不存在");
            }

            var userInfo = mapper.Map<UserInfoDto>(user);

            // 获取用户角色
            var userRoleIds = await dbContext.UserInRoles
                .Where(ur => ur.UserId == userId)
                .Select(ur => ur.RoleId)
                .ToListAsync();

            var userRoles = await dbContext.Roles
                .Where(r => userRoleIds.Contains(r.Id))
                .Select(r => r.Name)
                .ToListAsync();

            userInfo.Role = string.Join(',', userRoles);

            // 构建菜单
            var menus = BuildUserMenus(userRoles);

            return new UserMenuDto
            {
                User = userInfo,
                Menus = menus
            };
        }
        catch (Exception ex)
        {
            logger.LogError(ex, "获取用户菜单失败: UserId={UserId}", userContext.CurrentUserId);
            throw;
        }
    }

    /// <summary>
    /// 获取系统菜单结构（管理员权限）
    /// </summary>
    /// <returns>完整菜单结构</returns>
    [EndpointSummary("获取系统菜单结构")]
    [Authorize(Roles = "admin")]
    public async Task<List<MenuItemDto>> GetSystemMenusAsync()
    {
        try
        {
            await Task.CompletedTask; // 占位，实际可能从数据库读取
            return GetSystemMenuStructure();
        }
        catch (Exception ex)
        {
            logger.LogError(ex, "获取系统菜单失败");
            throw;
        }
    }

    /// <summary>
    /// 构建用户菜单
    /// </summary>
    /// <param name="userRoles">用户角色列表</param>
    /// <returns>菜单列表</returns>
    private List<MenuItemDto> BuildUserMenus(List<string> userRoles)
    {
        var allMenus = GetSystemMenuStructure();
        var userMenus = new List<MenuItemDto>();

        foreach (var menu in allMenus)
        {
            var filteredMenu = FilterMenuByRoles(menu, userRoles);
            if (filteredMenu != null)
            {
                userMenus.Add(filteredMenu);
            }
        }

        return userMenus;
    }

    /// <summary>
    /// 根据角色过滤菜单
    /// </summary>
    /// <param name="menu">菜单项</param>
    /// <param name="userRoles">用户角色</param>
    /// <returns>过滤后的菜单项</returns>
    private MenuItemDto? FilterMenuByRoles(MenuItemDto menu, List<string> userRoles)
    {
        // 检查当前菜单项是否有权限访问
        var hasAccess = !menu.RequiredRoles.Any() || menu.RequiredRoles.Any(role => userRoles.Contains(role));

        if (!hasAccess)
        {
            return null;
        }

        // 过滤子菜单
        var filteredChildren = new List<MenuItemDto>();
        foreach (var child in menu.Children)
        {
            var filteredChild = FilterMenuByRoles(child, userRoles);
            if (filteredChild != null)
            {
                filteredChildren.Add(filteredChild);
            }
        }

        // 如果当前菜单有权限但所有子菜单都没权限，仍然显示当前菜单
        return new MenuItemDto
        {
            Id = menu.Id,
            Name = menu.Name,
            Path = menu.Path,
            Icon = menu.Icon,
            Order = menu.Order,
            IsHidden = menu.IsHidden,
            RequiredRoles = menu.RequiredRoles,
            Children = filteredChildren
        };
    }

    /// <summary>
    /// 获取系统菜单结构定义
    /// </summary>
    /// <returns>菜单结构</returns>
    private List<MenuItemDto> GetSystemMenuStructure()
    {
        return new List<MenuItemDto>
        {
            // 首页
            new MenuItemDto
            {
                Id = "dashboard",
                Name = "首页",
                Path = "/",
                Icon = "dashboard",
                Order = 1,
                RequiredRoles = new List<string>()
            },
            
            // 仓库管理
            new MenuItemDto
            {
                Id = "repositories",
                Name = "仓库管理",
                Path = "/repositories",
                Icon = "repository",
                Order = 2,
                RequiredRoles = new List<string>(),
                Children = new List<MenuItemDto>
                {
                    new MenuItemDto
                    {
                        Id = "repository-list",
                        Name = "仓库列表",
                        Path = "/repositories",
                        Order = 1,
                        RequiredRoles = new List<string>()
                    },
                    new MenuItemDto
                    {
                        Id = "repository-create",
                        Name = "创建仓库",
                        Path = "/repositories/create",
                        Order = 2,
                        RequiredRoles = new List<string> { "admin" }
                    }
                }
            },
            
            // 用户管理
            new MenuItemDto
            {
                Id = "users",
                Name = "用户管理",
                Path = "/admin/users",
                Icon = "user",
                Order = 3,
                RequiredRoles = new List<string> { "admin" }
            },
            
            // 角色管理
            new MenuItemDto
            {
                Id = "roles",
                Name = "角色管理",
                Path = "/admin/roles",
                Icon = "role",
                Order = 4,
                RequiredRoles = new List<string> { "admin" }
            },
            
            // 权限管理
            new MenuItemDto
            {
                Id = "permissions",
                Name = "权限管理",
                Path = "/admin/permissions",
                Icon = "permission",
                Order = 5,
                RequiredRoles = new List<string> { "admin" },
                Children = new List<MenuItemDto>
                {
                    new MenuItemDto
                    {
                        Id = "role-permissions",
                        Name = "角色权限",
                        Path = "/admin/permissions/roles",
                        Order = 1,
                        RequiredRoles = new List<string> { "admin" }
                    },
                    new MenuItemDto
                    {
                        Id = "user-roles",
                        Name = "用户角色",
                        Path = "/admin/permissions/users",
                        Order = 2,
                        RequiredRoles = new List<string> { "admin" }
                    }
                }
            },
            
            // 系统设置
            new MenuItemDto
            {
                Id = "settings",
                Name = "系统设置",
                Path = "/admin/settings",
                Icon = "setting",
                Order = 6,
                RequiredRoles = new List<string> { "admin" },
                Children = new List<MenuItemDto>
                {
                    new MenuItemDto
                    {
                        Id = "general-settings",
                        Name = "基本设置",
                        Path = "/admin/settings/general",
                        Order = 1,
                        RequiredRoles = new List<string> { "admin" }
                    },
                    new MenuItemDto
                    {
                        Id = "system-info",
                        Name = "系统信息",
                        Path = "/admin/settings/system",
                        Order = 2,
                        RequiredRoles = new List<string> { "admin" }
                    }
                }
            },
            
            // 微调管理
            new MenuItemDto
            {
                Id = "finetune",
                Name = "微调管理",
                Path = "/admin/finetune",
                Icon = "finetune",
                Order = 7,
                RequiredRoles = new List<string> { "admin" }
            },
            
            // 统计分析
            new MenuItemDto
            {
                Id = "statistics",
                Name = "统计分析",
                Path = "/admin/statistics",
                Icon = "statistics",
                Order = 8,
                RequiredRoles = new List<string> { "admin" }
            },
            
            // 个人设置
            new MenuItemDto
            {
                Id = "profile",
                Name = "个人设置",
                Path = "/settings",
                Icon = "profile",
                Order = 99,
                RequiredRoles = new List<string>()
            }
        };
    }

    /// <summary>
    /// 检查用户是否有访问指定路径的权限
    /// </summary>
    /// <param name="path">路径</param>
    /// <param name="userId">用户ID</param>
    /// <returns>是否有权限</returns>
    [EndpointSummary("检查用户路径访问权限")]
    [AllowAnonymous]
    public async Task<bool> CheckUserPathPermissionAsync(string path, string? userId = null)
    {
        try
        {
            userId ??= userContext.CurrentUserId;
            
            if (string.IsNullOrEmpty(userId))
            {
                return false;
            }

            // 获取用户角色
            var userRoleIds = await dbContext.UserInRoles
                .Where(ur => ur.UserId == userId)
                .Select(ur => ur.RoleId)
                .ToListAsync();

            var userRoles = await dbContext.Roles
                .Where(r => userRoleIds.Contains(r.Id))
                .Select(r => r.Name)
                .ToListAsync();

            // 查找路径对应的菜单项
            var allMenus = GetSystemMenuStructure();
            var menuItem = FindMenuByPath(allMenus, path);

            if (menuItem == null)
            {
                // 如果路径不在菜单定义中，默认允许访问（比如动态路由）
                return true;
            }

            // 检查权限
            return !menuItem.RequiredRoles.Any() || menuItem.RequiredRoles.Any(role => userRoles.Contains(role));
        }
        catch (Exception ex)
        {
            logger.LogError(ex, "检查用户路径权限失败: Path={Path}, UserId={UserId}", path, userId);
            return false;
        }
    }

    /// <summary>
    /// 根据路径查找菜单项
    /// </summary>
    /// <param name="menus">菜单列表</param>
    /// <param name="path">路径</param>
    /// <returns>菜单项</returns>
    private MenuItemDto? FindMenuByPath(List<MenuItemDto> menus, string path)
    {
        foreach (var menu in menus)
        {
            if (menu.Path == path)
            {
                return menu;
            }

            var childResult = FindMenuByPath(menu.Children, path);
            if (childResult != null)
            {
                return childResult;
            }
        }

        return null;
    }
} 

File: web/app/services/menuService.ts
'use client';
import { fetchApi, API_URL, ApiResponse } from './api';
import { UserInfo } from './roleService';

// 菜单相关接口定义
export interface MenuItem {
  id: string;
  name: string;
  path: string;
  icon?: string;
  order: number;
  isHidden: boolean;
  requiredRoles: string[];
  children: MenuItem[];
}

export interface UserMenu {
  user?: UserInfo;
  menus: MenuItem[];
}

// 菜单管理API服务
export class MenuService {
  private baseUrl = `${API_URL}/api/Menu`;

  // 获取当前用户的菜单
  async getUserMenu(): Promise<UserMenu> {
    const {data} = await fetchApi<any>(`${this.baseUrl}/UserMenu`, {
      method: 'GET',
    });

    if (data.code !== 200) {
      throw new Error(data.message || '获取用户菜单失败');
    }
    return data.data!;
  }

  // 获取系统菜单结构（管理员权限）
  async getSystemMenus(): Promise<MenuItem[]> {
    const result = await fetchApi<MenuItem[]>(`${this.baseUrl}/SystemMenu`, {
      method: 'GET',
    });

    if (!result.success) {
      throw new Error(result.error || '获取系统菜单失败');
    }

    return result.data!;
  }

  // 检查用户路径访问权限
  async checkUserPathPermission(path: string, userId?: string): Promise<boolean> {
    const params = new URLSearchParams({ path });
    if (userId) params.append('userId', userId);

    const { data } = await fetchApi<any>(`${this.baseUrl}/CheckUserPathPermission?${params}`, {
      method: 'POST',
    });

    if (data.code !== 200) {
      throw new Error(data.message || '检查用户路径权限失败');
    }

    return data.data!;
  }

  // 本地菜单工具方法

  // 根据路径查找菜单项
  findMenuByPath(menus: MenuItem[], path: string): MenuItem | null {
    for (const menu of menus) {
      if (menu.path === path) {
        return menu;
      }

      const childResult = this.findMenuByPath(menu.children, path);
      if (childResult) {
        return childResult;
      }
    }
    return null;
  }

  // 根据角色过滤菜单
  filterMenusByRoles(menus: MenuItem[], userRoles: string[]): MenuItem[] {
    const filteredMenus: MenuItem[] = [];

    for (const menu of menus) {
      // 检查当前菜单项是否有权限访问
      const hasAccess = !menu.requiredRoles.length || menu.requiredRoles.some(role => userRoles.includes(role));

      if (hasAccess) {
        // 递归过滤子菜单
        const filteredChildren = this.filterMenusByRoles(menu.children, userRoles);

        // 创建过滤后的菜单项
        const filteredMenu: MenuItem = {
          ...menu,
          children: filteredChildren
        };

        filteredMenus.push(filteredMenu);
      }
    }

    return filteredMenus;
  }

  // 扁平化菜单结构
  flattenMenus(menus: MenuItem[]): MenuItem[] {
    const flattened: MenuItem[] = [];

    function flatten(items: MenuItem[]) {
      for (const item of items) {
        flattened.push(item);
        if (item.children.length > 0) {
          flatten(item.children);
        }
      }
    }

    flatten(menus);
    return flattened;
  }

  // 获取面包屑路径
  getBreadcrumb(menus: MenuItem[], currentPath: string): MenuItem[] {
    const breadcrumb: MenuItem[] = [];

    function findPath(items: MenuItem[], path: string, current: MenuItem[]): boolean {
      for (const item of items) {
        const newPath = [...current, item];

        if (item.path === path) {
          breadcrumb.push(...newPath);
          return true;
        }

        if (item.children.length > 0) {
          if (findPath(item.children, path, newPath)) {
            return true;
          }
        }
      }
      return false;
    }

    findPath(menus, currentPath, []);
    return breadcrumb;
  }

  // 检查菜单是否激活（当前路径或子路径）
  isMenuActive(menu: MenuItem, currentPath: string): boolean {
    if (menu.path === currentPath) {
      return true;
    }

    // 检查是否有子菜单激活
    return menu.children.some(child => this.isMenuActive(child, currentPath));
  }

  // 获取管理员菜单列表（用于权限配置）
  getAdminMenuList(): MenuItem[] {
    return [
      {
        id: 'admin',
        name: '系统管理',
        path: '/admin',
        icon: 'admin',
        order: 1,
        isHidden: false,
        requiredRoles: ['admin'],
        children: [
          {
            id: 'admin-users',
            name: '用户管理',
            path: '/admin/users',
            icon: 'user',
            order: 1,
            isHidden: false,
            requiredRoles: ['admin'],
            children: []
          },
          {
            id: 'admin-roles',
            name: '角色管理',
            path: '/admin/roles',
            icon: 'role',
            order: 2,
            isHidden: false,
            requiredRoles: ['admin'],
            children: []
          },
          {
            id: 'admin-permissions',
            name: '权限管理',
            path: '/admin/permissions',
            icon: 'permission',
            order: 3,
            isHidden: false,
            requiredRoles: ['admin'],
            children: [
              {
                id: 'admin-role-permissions',
                name: '角色权限',
                path: '/admin/permissions/roles',
                order: 1,
                isHidden: false,
                requiredRoles: ['admin'],
                children: []
              },
              {
                id: 'admin-user-roles',
                name: '用户角色',
                path: '/admin/permissions/users',
                order: 2,
                isHidden: false,
                requiredRoles: ['admin'],
                children: []
              }
            ]
          },
          {
            id: 'admin-repositories',
            name: '仓库管理',
            path: '/admin/repositories',
            icon: 'repository',
            order: 4,
            isHidden: false,
            requiredRoles: ['admin'],
            children: []
          },
          {
            id: 'admin-settings',
            name: '系统设置',
            path: '/admin/settings',
            icon: 'setting',
            order: 5,
            isHidden: false,
            requiredRoles: ['admin'],
            children: []
          }
        ]
      }
    ];
  }
}

// 导出服务实例
export const menuService = new MenuService(); 